### DISCUSSION TITLE
Generate unittes XML Junit reports

### DISCUSSION THREAD

**Discussion by madecsmile** [OCA/odoo-community.org#167] [Ideas]
Hi, 
As far as I know, Odoo tests does not generate tests reports. 

I made a small module wich use https://pypi.org/project/unittest-xml-reporting/ generate reports. 
The override is simple, 2 lines on new_run_suite method.

Then, the tests results are fully integrated on the CI (I am using Gitlab, it should do the same on  Github).

Does that module can be published on OCA repo ? 

Thanks, 
Madec

**Comment by pedrobaeza** [comment]
It can be interesting, although the override you are mentioning seems that you are touching Odoo codebase. Is that true? Can you share a module link?

**Reply by madecsmile** [reply]
Hi, 
Here it is : https://github.com/madecsmile/odoo_test_xmlrunner

Martin

**Reply by pedrobaeza** [reply]
Thanks, I see you totally replace the method `new_run_suite`. It's not likely that this method changes, but is it possible to do a less aggressive approach calling at least super with the proper replacements?

And once this module is put, how this affect the results on Gitlab/GitHub?

**Reply by madecsmile** [reply]
Maybe I can have a less aggressive override, I need some times to achieve this. 


Once this module is put, with some CI confgiration (to be documented), the report is visibles on Gitlab (or Github) job, like that screen shot :
![pipelines_junit_test_report_v13_10](https://github.com/OCA/odoo-community.org/assets/66998705/95b23f0c-22eb-4d3f-a9cb-c42528806771)

Test results can be expended to see logs of each tests, error details durations.

**Reply by madecsmile** [reply]
@pedrobaeza so far if I do not replace that method, I need to replace an other one in a dirty way too (TestSuite.run). Those codes are not meant to me overridden, that doesn't surprise me.

**Reply by pedrobaeza** [reply]
OK, I see. Then this the less bad one. Let's see if Odoo contemplates this for newer versions. Please prepare the OCA module for the repo server-tools, and please include the samples for GitHub/Gitlab.

**Reply by yajo** [reply]
I'm seeing your code and it looks really good to me. It's probably easier to maintain than keeping https://github.com/odoo/odoo/pull/151728 up to date as I try to do now. Of course, upstream support would be the best option, but having this module available should be good.

TBH I'd love it much more if pytest-odoo supported the canonical way of testing addons (see https://github.com/camptocamp/pytest-odoo/issues/67). But that isn't the case right now and I don't know how hard it would be to implement.

**Reply by madecsmile** [reply]
Hello @pedrobaeza, you will find my PR here : https://github.com/OCA/server-tools/pull/2957
Best regards.

**Comment by ap-wtioit** [comment]
You mention that your are using gitlab, how would i test this out on our instance?

We are currently running the tests with pytest (in doodba):
```bash
/usr/local/bin/pytest --odoo-database="${PGDATABASE}" --odoo-config=/opt/odoo/auto/odoo_test.conf --odoo-http --junitxml=/opt/odoo/auto/test-artifacts/junit.xml "${pytest_args[@]}"
```

**Reply by madecsmile** [reply]
Here is the command line I use to generate coverage and test results.

`coverage run --rcfile=config/.coveragerc odoo/odoo-bin -c config/odoo-dev.conf --test-tags "$TEST_TAGS" --stop-after-init`

The results are created in the dir specified in test_result_directory configuration.

**Comment by ap-wtioit** [comment]
And we should also have a look at https://github.com/odoo/odoo/pull/151728 where @yajo made an effort to bring xml reports to Odoo

**Reply by yajo** [reply]
I published that but sadly I have to admit it's probably doomed for oblivion at upstream's eyes.

**Comment by matteoopenf** [comment]
https://github.com/OCA/server-tools/issues/3156
